drivers/storage/port/scsiport/scsi.c

```diff
@@ -683,8 +686,21 @@ SpiProcessCompletedRequest(
     /* Flush adapter if needed */
     if (SrbInfo->BaseOfMapRegister)
     {
-        /* TODO: Implement */
-        ASSERT(FALSE);
+        for(int i = 0;
+            i < SrbInfo->NumberOfMapRegisters     &&   /* Prevent Buffer overrun */
+            transferLen < Srb->DataTransferLength &&   /* Stop if request is complete */
+            pListIter;                                 /* Valid Pointer Check */
+            pListIter++, i++)
```

Comment:
(Temporary suggestion before following remark is addressed):
```suggestion
        ULONG i;

        /* Calculate size of all pending data in the registers,
         * that can be held in the data buffer */
        for (i = 0;
             i < SrbInfo->NumberOfMapRegisters     &&
             transferLen < Srb->DataTransferLength &&
             pListIter;                   /* Valid Pointer Check */
             i++, pListIter++)
```
Remark: I think the "valid pointer check" could be moved *outside* the for-loop: `pListIter` is initialized to the value of `Srb->ScatterGather` that is either NULL or not-NULL already, so one single check can verify that only once before starting the loop.

Reply:
I tend to think the whole loop could benefit from a little simplification. What would you think of something like this?
```
// if this is null, do we really need to call IoFlushAdapterBuffers? If not, this can be combined with the BaseOfMapRegister check above
if (SrbInfo->ScatterGather != NULL)
{
    for (ULONG i = 0; i < SrbInfo->NumberOfMapRegisters; i++)
    {
        if (transferLen >= Srb->DataTransferLength)
        {
            /* Request is complete */
            break;
        }
        transferLen += SrbInfo->ScatterGather[i].Length;
    }
}

```

Reply:
You're right, that is much better. Fixed